### PR TITLE
`pyright` fixes

### PR DESCRIPTION
**Pull request by DanielYang59**
## Summary

- `pyright` fixes, a follow up PR of #3757.
- Enable `pyright` in linting workflow and `pre-commit`.

### CODE REVIEW

File: pymatgen/core/interface.py [hunk 1]
```diff
@@ -2160,7 +2163,9 @@ def slab_from_csl(csl, surface, normal, trans_cry, max_search=20, quick_gen=Fals
                     ab_norm = ab_norm_temp
                     t_matrix[0] = ii[0]
                     t_matrix[1] = ii[1]
-                elif area_temp < area or (abs(area - area_temp) < 1.0e-8 and ab_norm_temp < ab_norm):
+
+                # DEBUG (@DanielYang59): ab_norm would certainly be unbound
+                elif area_temp < area or (abs(area - area_temp) < 1.0e-8 and ab_norm_temp < ab_norm):  # type: ignore  # temporary supress
```
Comment (DanielYang59) [inline pymatgen/core/interface.py:2168]: Seems a bug here, `ab_norm` would certainly be unbound:
https://github.com/materialsproject/pymatgen/blob/ad6eafee2474718813cba57aa8e1752f1181b639/pymatgen/core/interface.py#L2153-L2167

File: pymatgen/analysis/graphs.py [hunk 2]
```diff
@@ -1489,6 +1490,8 @@ def get_subgraphs_as_molecules(self, use_weights: bool = False) -> list[Molecule
         # without adding extra logic
         if getattr(self, "_supercell_sg", None) is None:
             self._supercell_sg = supercell_sg = self * (3, 3, 3)
+        else:  # TODO: need double check
+            raise RuntimeError("Supercell spacegroup is not None.")
```
Comment (DanielYang59) [inline pymatgen/analysis/graphs.py:1494]: This condition is weird, need some expert advice on this, this function only works when `self._supercell_sg` is `None`:

https://github.com/materialsproject/pymatgen/blob/ad6eafee2474718813cba57aa8e1752f1181b639/pymatgen/analysis/graphs.py#L1467-L1544

**Comment by DanielYang59** [comment]
@janosh Please review this, all `pyright` errors are fixed :) Not sure why `pre-commit` is failing though (passing locally).

**Review by janosh — COMMENTED** [review]
this is great, thanks so much! @ml-evs will be happy to hear this :)

i think `pyright` is failing in pre-commit CI because it can't find the dependencies installed into a virtual env. try adding to `pyproject.toml` (CC @njzjz)

```toml
[tool.pyright]
venv = "pmg"
venvPath = "~/micromamba/envs/pmg" # TODO check this path is correct, not sure where micromamba puts envs
```

**Comment by DanielYang59** [comment]
Thanks for the input. I'll have a try (not sure where `micromamba` put venv too... Never use it). 

https://github.com/materialsproject/pymatgen/pull/3777#discussion_r1574080211 This has to be confirmed :)

**Comment by janosh** [comment]
sorry, actually this was a stupid suggestion. i thought `pyright` was failing in GitHub action but actually it's only failing in pre-commit CI which runs in a completely different container where `pymatgen` are not installed anyway. we could [manually install all of them](https://pre-commit.com/#config-additional_dependencies) but that would be tedious to maintain as deps change. better to only run `pyright` in GitHub Action and `skip` it in pre-commit CI imo

**Comment by DanielYang59** [comment]
It's weird, `pyright` still failed. But I think `pyright` worked fine in #3757? Back that time we already have `micromamba` installed in #3729?

In my local test machine, the path is `/home/yang/micromamba/envs/pmg` seems correct.

File: pymatgen/analysis/graphs.py (hunk 2 above)
Comment (DanielYang59) [reply pymatgen/analysis/graphs.py:1494]: I just looks into the history, it was added by @mkhorton in #831 (line 937), and the original code is written this way. Maybe we should keep it as is for now?

**Comment by njzjz** [comment]
> Not sure why `pre-commit` is failing though (passing locally), also tried to downgrade to previous but still no luck. Noticed a similar report [RobertCraigie/pyright-python#173](https://github.com/RobertCraigie/pyright-python/issues/173), but not replied.

I think the reason is that pre-commit.ci disallow the network at the runtime (see https://github.com/pre-commit-ci/issues/issues/196#issuecomment-1810937079), but pyright-python only downloads pyright at the first runtime (see https://github.com/RobertCraigie/pyright-python/issues/225).

**Comment by janosh** [comment]
thanks @njzjz! that's good to know. more reason to skip it in pre-commit CI.

@DanielYang59 could you add `pyright` here?

https://github.com/materialsproject/pymatgen/blob/ad6eafee2474718813cba57aa8e1752f1181b639/.pre-commit-config.yaml#L5

**Comment by DanielYang59** [comment]
Thanks @njzjz! It's very helpful.

**Comment by DanielYang59** [comment]
@janosh. Yes it makes sense, we may not need `pyright` on both CI and pre-commit. Skipped in pre-commit, thanks!

**Comment by DanielYang59** [comment]
Ah, wait! I'm pretty sure I added a timeout of 60 seconds for `request` to prevent it from hanging forever, but it seems 60 seconds it insufficient?

```
self = <ext.test_cod.TestCOD testMethod=test_get_structure_by_id>

    def test_get_structure_by_id(self):
>       struct = COD().get_structure_by_id(2002926)

tests/ext/test_cod.py:31: 
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ 
pymatgen/ext/cod.py:93: in get_structure_by_id
    response = requests.get(f"http://{self.url}/cod/{cod_id}.cif", timeout=60)
../../../micromamba/envs/pmg/lib/python3.11/site-packages/requests/api.py:73: in get
    return request("get", url, params=params, **kwargs)
../../../micromamba/envs/pmg/lib/python3.11/site-packages/requests/api.py:59: in request
    return session.request(method=method, url=url, **kwargs)
../../../micromamba/envs/pmg/lib/python3.11/site-packages/requests/sessions.py:589: in request
    resp = self.send(prep, **send_kwargs)
../../../micromamba/envs/pmg/lib/python3.11/site-packages/requests/sessions.py:703: in send
    r = adapter.send(request, **kwargs)
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ 

self = <requests.adapters.HTTPAdapter object at 0x7f3737f1e250>
request = <PreparedRequest [GET]>, stream = False
timeout = Timeout(connect=60, read=60, total=None), verify = True, cert = None
proxies = OrderedDict()
```

**Comment by janosh** [comment]
feel free to skip the COD test as well. pinging live APIs in CI is bad practice anyway

**Comment by DanielYang59** [comment]
Thanks for reviewing @janosh.

Also ping @ml-evs: all `pyright` errors have been fixed :)

**Comment by DanielYang59** [comment]
@janosh. I notice [CI is failing](https://github.com/materialsproject/pymatgen/actions/runs/8793843655/job/24132330303) after this PR. Did we actually break anything?

**Comment by janosh** [comment]
thanks for the quick fix!

**Comment by janosh** [comment]
btw, very happy that we can enable `pyright`'s possibly unbound variable detection across the code base thanks to this PR. that will really help automated code quality assurance

**Comment by DanielYang59** [comment]
> btw, very happy that we can enable `pyright`'s possibly unbound variable detection across the code base thanks to this PR. that will really help automated code quality assurance

Great to know that! And very glad I could help :)

**Pull request merged**